File: expanding-bot.js
const CONSTANTS = require('./helper-package-js/constants');
const { ACTION, FACE, FLOWER, MOVE } = require('./helper-package-js/constants');

const { Action } = require('./helper-package-js/action');
const { Bot } = require('./helper-package-js/bot');
const { Position } = require('./helper-package-js/position');

const BEES_TO_HIVE_RATIO = 30;

class ExpandingBot extends Bot {
    constructor() {
        super("ExpandingBot");
    }

    getActions() {
        const actions = [];

        for (let beeCell of this.beeCells) {
            const bee = beeCell.bee;
            const pos = beeCell.pos;
            if (bee.pollen >= bee.count) {
                // If the bees have pollen, move to closest queen or hive.
                if (!this.isBesideHiveOrQueen(pos)) {
                    let minPath = this.getMinPath(pos, this.queenBee.pos);
                    for (let hiveCell of this.hiveCells) {
                        let path = this.getMinPath(pos, hiveCell.pos);
                        if (path.distance < minPath.distance) {
                            minPath = path;
                        }
                    }
                    actions.push(new Action({
                        type: ACTION.MOVE,
                        pos: pos,
                        move: minPath.move,
                        face: minPath.move - 1 /* Faces move direction. */,
                    }));
                }
            } else {
                // Move to closest flower to collect pollen.
                const firstFlower = this.flowerCells[0];
                let minPath = this.getMinPath(pos, firstFlower.pos);
                for (let i = 1; i < this.flowerCells.length; ++i) {
                    const path = this.getMinPath(pos, this.flowerCells[i].pos);
                    if (path.distance < minPath.distance) {
                        minPath = path;
                    }
                }
                actions.push(new Action({
                    type: ACTION.MOVE,
                    pos: pos,
                    move: minPath.move,
                    face: minPath.move - 1 /* Faces move direction. */,
                }));
            }
        }

        const queenPos = this.queenBee.pos;
        let pollenNotUsed = this.queenBee.pollen;
        let posPotency = this.currMap.map[queenPos.y][queenPos.x].potency;

        // Create hive if the bees to hive ratio is higher than what we want.
        if (this.totalBees > BEES_TO_HIVE_RATIO * this.hiveCells.length && posPotency == FLOWER.NONE) {
            pollenNotUsed -= CONSTANTS.HIVE_POLLEN_AMOUNT;
            actions.push(new Action({
                type: ACTION.CREATE_HIVE,
            }));
        }

        // If queen is on a hive or flower, move to an open spot.
        if (!this.isEmpty(queenPos)) {
            if (this.isEmpty(this.getBoundedPos(queenPos.x + 1, queenPos.y))) {
                actions.push(new Action({
                    type: ACTION.MOVE_QUEEN,
                    move: MOVE.RIGHT,
                    face: FACE.RIGHT,
                }));
            } else if (this.isEmpty(this.getBoundedPos(queenPos.x, queenPos.y + 1))) {
                actions.push(new Action({
                    type: ACTION.MOVE_QUEEN,
                    move: MOVE.DOWN,
                    face: FACE.DOWN,
                }));
            } else if (this.isEmpty(this.getBoundedPos(queenPos.x - 1, queenPos.y))) {
                actions.push(new Action({
                    type: ACTION.MOVE_QUEEN,
                    move: MOVE.LEFT,
                    face: FACE.LEFT,
                }));
            } else if (this.isEmpty(this.getBoundedPos(queenPos.x, queenPos.y - 1))) {
                actions.push(new Action({
                    type: ACTION.MOVE_QUEEN,
                    move: MOVE.UP,
                    face: FACE.UP,
                }));
            } else {
                // Everywhere is blocked, so just move randomly.
                const move = Math.floor(Math.random() * 4) + 1;
                actions.push(new Action({
                    type: ACTION.MOVE_QUEEN,
                    move: move,
                    face: move - 1 /* Faces move direction. */,
                }));
            }
        }

        if (this.hiveCells.length > 0) {
            const totalBeesPerHive =
                Math.floor(Math.floor(pollenNotUsed / CONSTANTS.BEE_POLLEN_AMOUNT) / this.hiveCells.length);
            for (let hiveCell of this.hiveCells) {
                // Makes sure we don't overspawn bees.
                if (totalBeesPerHive > CONSTANTS.MAX_BEES) {
                    actions.push(new Action({
                        type: ACTION.SPAWN,
                        pos: hiveCell.pos,
                        amount: CONSTANTS.MAX_BEES,
                    }));
                } else {
                    actions.push(new Action({
                        type: ACTION.SPAWN,
                        pos: hiveCell.pos,
                        amount: totalBeesPerHive,
                    }));
                }
            }
        }

        return actions;
    }

    isEmpty(pos) {
        return this.currMap.map[pos.y][pos.x].potency == FLOWER.NONE &&
            this.currMap.map[pos.y][pos.x].ownerId == CONSTANTS.NEUTRAL_ID;
    }
}

(async function main() {
    const expandingBot = new ExpandingBot();
    await expandingBot.run();
})();

File: expanding_bot.py
from helper_package_python.constants import (
	ActionType, 
	Move, 
	Face, 
	Flower, 
	HIVE_POLLEN_AMOUNT, 
	BEE_POLLEN_AMOUNT,
	MAX_BEES,
	NEUTRAL_ID
)

from helper_package_python.bot import Bot
from helper_package_python.action import Action

from random import randint
import math

BEES_TO_HIVE_RATIO = 30

class ExpandingBot(Bot):
	def __init__(self):
		super().__init__("ExpandingBot")

	def getActions(self):
		actions = []

		for beeCell in self.beeCells:
			bee = beeCell.bee
			pos = beeCell.pos

			if bee.pollen >= bee.count:
				# If the bees have pollen, move to queen.
				if not self.isBesideHiveOrQueen(pos):
					minPath = self.getMinPath(pos, self.queenBee.pos)

					for hiveCell in self.hiveCells:
						path = self.getMinPath(pos, hiveCell.pos)
						if path.distance < minPath.distance:
							minPath = path

					actions.append(Action(
						ActionType.MOVE,
						pos,
						minPath.move,
						Face(minPath.move - 1) # Face move direction
					))
			else:
				# Move to closest flower to collect pollen.
				firstFlower = self.flowerCells[0]
				minPath = self.getMinPath(pos, firstFlower.pos)
				
				for i in range(1, len(self.flowerCells)):
					path = self.getMinPath(pos, self.flowerCells[i].pos)
					if path.distance < minPath.distance:
						minPath = path

				actions.append(Action(
					ActionType.MOVE,
					pos,
					minPath.move,
					Face(minPath.move - 1) # Face move direction
				))

		queenPos = self.queenBee.pos
		pollenNotUsed = self.queenBee.pollen
		posPotency = self.currMap.map[queenPos.y][queenPos.x].potency

		# Create hive if the bees to hive ratio is higer than what we want
		if (self.totalBees > BEES_TO_HIVE_RATIO * len(self.hiveCells)) and (posPotency == Flower.NONE):
			pollenNotUsed -= HIVE_POLLEN_AMOUNT
			actions.append(Action(
				ActionType.CREATE_HIVE
			))

		# If queen is on hive or flower, move to an open spot.
		if not self.isEmpty(queenPos):
			if self.isEmpty(self.getBoundedPos(queenPos.x + 1, queenPos.y)):
				actions.append(Action(
					ActionType.MOVE_QUEEN,
					None,
					Move.RIGHT,
					Face.RIGHT
				))
			elif self.isEmpty(self.getBoundedPos(queenPos.x, queenPos.y + 1)):
				actions.append(Action(
					ActionType.MOVE_QUEEN,
					None,
					Move.DOWN,
					Face.DOWN
				))
			elif self.isEmpty(self.getBoundedPos(queenPos.x - 1, queenPos.y)):
				actions.append(Action(
					ActionType.MOVE_QUEEN,
					None,
					Move.LEFT,
					Face.LEFT
				))
			elif self.isEmpty(self.getBoundedPos(queenPos.x, queenPos.y - 1)):
				actions.append(Action(
					ActionType.MOVE_QUEEN,
					None,
					Move.UP,
					Face.UP
				))
			else:
				# Move randomly
				move = randint(1, 4)
				actions.append(Action(
					ActionType.MOVE_QUEEN,
					None,
					Move(move),
					Face(move - 1) # Faces move direction
				))

		if not self.hiveCells:
			totalBeesPerHive = math.floor(math.floor(pollenNotUsed / BEE_POLLEN_AMOUNT) / len(self.hiveCells))
			for hiveCell in self.hiveCells:
				# Makes sure we dont overspawn bees
				if totalBeesPerHive > MAX_BEES:
					actions.append(Action(
						ActionType.SPAWN,
						hiveCell.pos,
						MAX_BEES
					))
				else:
					actions.append(Action(
						ActionType.SPAWN,
						hiveCell.pos,
						totalBeesPerHive
					))

		return actions

	def isEmpty(self, pos):
		return self.currMap.map[pos.y][pos.x].potency == Flower.NONE and \
            self.currMap.map[pos.y][pos.x].ownerId == NEUTRAL_ID

def main():
	expandingBot = ExpandingBot()
	expandingBot.run()

main()